src/main/scala/com/microsoft/hyperspace/actions/OptimizeAction.scala

```diff
@@ -106,20 +106,38 @@ class OptimizeAction(
 
     if (filesToOptimize.isEmpty) {
       throw NoChangesException(
-        "Optimize aborted as no index files smaller than " +
+        "Optimize aborted as no optimizable index files smaller than " +
           s"${HyperspaceConf.optimizeFileSizeThreshold(spark)} found.")
     }
   }
 
   private lazy val (filesToOptimize, filesToIgnore): (Seq[FileInfo], Seq[FileInfo]) = {
-    if (mode.equalsIgnoreCase(IndexConstants.OPTIMIZE_MODE_QUICK)) {
-      val threshold = HyperspaceConf.optimizeFileSizeThreshold(spark)
-      previousIndexLogEntry.content.fileInfos.toSeq.partition(_.size < threshold)
-    } else {
-      // For 'full' mode, put all the existing index files into 'filesToOptimize' partition so
-      // that one file is created per bucket.
-      (previousIndexLogEntry.content.fileInfos.toSeq, Seq())
+    val (possibleCandidates, largeFilesToIgnore) =
+      if (mode.equalsIgnoreCase(IndexConstants.OPTIMIZE_MODE_QUICK)) {
+        val threshold = HyperspaceConf.optimizeFileSizeThreshold(spark)
+        previousIndexLogEntry.content.fileInfos.toSeq.partition(_.size < threshold)
+      } else {
+        // For 'full' mode, put all the existing index files into 'filesToOptimize' partition
+        // so that one file is created per bucket.
+        (previousIndexLogEntry.content.fileInfos.toSeq, Seq())
+      }
+
+    // Check if there is only one index file in a bucket and exclude the index file
+    // as it doesn't need to be optimized.
+    val filesPerBucket = mutable.Map[Int, List[FileInfo]]()
+    possibleCandidates.foreach { f =>
+      val key = BucketingUtils.getBucketId(new Path(f.name).getName).get
+      if (!filesPerBucket.contains(key)) {
+        filesPerBucket.put(key, List(f))
+      } else {
+        filesPerBucket.update(key, filesPerBucket(key) :+ f)
+      }
     }
+
+    // Get candidate index files for optimization based on the length of file list per bucket.
```

Comment:
nit: How about explaining why:
```
// Files belonging to buckets that have only one file are ignored because they are already optimized.
```
or you can explain the opposite (e.g., multiple files within a bucket can be merged).